cognitedata/reveal
# feat(react-components): CoreDM queries

#### Type of change

![Feat](https://img.shields.io/badge/Type-Feat-green)  

#### Jira ticket :blue_book:


https://cognitedata.atlassian.net/browse/BND3D-4541

## Description :pencil:


Now - this was originally just a PR to implement the CoreDM-part of the Fdm3dDataProvider. However, a few cascading effects occured, making this monstrosity of a PR. I'll go through the relevant changes:

- Introduce typed queries. This is adapted directly from @christjt 's implementation in fusion: https://github.com/cognitedata/fusion/blob/master/libs/3d/src/api/datamodeling/sdk-extensions/queryNodesAndEdges.ts
  - This of course had a series of effects, due to requiring more explicitly typed calls. This especially lead to a long list of necessary changes in the Scenes-implementation, where I ended up doing a major refactoring to make the code more maintainable and usable type-wise.
- The `FdmCadConnection` type no longer uses nodeIds, but treeIndex. This type is used as a common grounds for an "asset mapping" for the legacy FDM-API and the new CoreDM. Since CoreDM does not really have a concept of nodeId, I decided to change it to `treeIndex` for future proofing. This may have performance implications for the legacy code, as I was forced to make another query translating between nodeIds and treeIndices. But all the better for CoreDM.
- Along with this, the caches now use treeIndex to as cache key, instead of nodeId, which also lead to a few changes.

I'll do my best to delegate the review across multiple team members

Remaining work: 
- ~~Implement remaining functions (3-4 of them?)~~
- ~~Have proper data~~
- ~~Test it all~~



## How has this been tested? :mag:



In Fusion using this PR: https://github.com/cognitedata/fusion/pull/7854

## Test instructions :information_source:



- Open in Fusion using PR mentioned above, enter project 3d-dev-aws
- In ThreeDSplitView.tsx (in fusion) and in ThreeDWidget, make sure that the `RevealContext` has the `useCoreDm` prop set to `true`
- See that all 3D functionality (listing, search, clicking in list, clicking in 3D, opening instance page, show mappings etc.) works


## Checklist :ballot_box_with_check:
